src/libmongoc/src/mongoc/mongoc-topology-background-monitoring.c

```diff
@@ -0,0 +1,838 @@
+ * Caller must hold topology lock.
+ */
+static bool
+_server_monitor_try_shutdown_and_destroy (
+   mongoc_server_monitor_t *server_monitor)
+{
+   bool is_shutdown;
+
+   bson_mutex_lock (&server_monitor->shared.mutex);
+   is_shutdown = server_monitor->shared.is_shutdown;
+   server_monitor->shared.shutting_down = true;
+   mongoc_cond_signal (&server_monitor->shared.cond);
+   bson_mutex_unlock (&server_monitor->shared.mutex);
+
+   /* If the server monitor thread has exited, join so the internal thread state
+    * can be freed. Joining can only be done once the server monitor has
+    * completed shutdown. Otherwise, the server monitor may be in the middle of
+    * scanning (and therefore may need to take the topology mutex again).
+    *
+    * Since the topology mutex is locked, we're guaranteed that only one thread
+    * will join.
+    */
+   if (is_shutdown) {
+      TRACE ("sm (%d) try join start", server_monitor->server_id);
+      COMMON_PREFIX (thread_join) (server_monitor->thread);
+      TRACE ("sm (%d) try join end", server_monitor->server_id);
+      _server_monitor_destroy (server_monitor);
+      return true;
+   }
+   return false; /* Still waiting for shutdown. */
+}
+
+/* Request scan of a single server.
+ *
+ * Caller does not need to have topology mutex locked.
+ * Locks server_monitor mutex to deliver scan_requested.
+ */
+static void
+_server_monitor_request_scan (mongoc_server_monitor_t *server_monitor)
+{
+   bson_mutex_lock (&server_monitor->shared.mutex);
+   server_monitor->shared.scan_requested = true;
+   mongoc_cond_signal (&server_monitor->shared.cond);
+   bson_mutex_unlock (&server_monitor->shared.mutex);
+}
+
+/* Create a server monitor if necessary.
+ *
+ * Called by monitor threads and application threads when reconciling the
+ * topology description. Caller must have topology mutex locked.
+ */
+static void
+_background_monitor_reconcile_server_monitor (mongoc_topology_t *topology,
+                                              mongoc_server_description_t *sd)
+{
+   mongoc_set_t *server_monitors;
+   mongoc_server_monitor_t *server_monitor;
+
+   server_monitors = topology->server_monitors;
+   server_monitor = mongoc_set_get (server_monitors, sd->id);
+
+   if (server_monitor) {
+      return;
+   }
+
+   /* Add a new server monitor. */
+   server_monitor = bson_malloc0 (sizeof (*server_monitor));
+   server_monitor->server_id = sd->id;
+   memcpy (&server_monitor->host, &sd->host, sizeof (mongoc_host_list_t));
+   server_monitor->topology = topology;
+   server_monitor->heartbeat_frequency_ms =
+      topology->description.heartbeat_msec;
+   server_monitor->min_heartbeat_frequency_ms =
+      topology->min_heartbeat_frequency_msec;
+   server_monitor->connect_timeout_ms = topology->connect_timeout_msec;
+   server_monitor->uri = mongoc_uri_copy (topology->uri);
+/* TODO: Do not retrieve ssl opts from topology scanner. They should be stored
```

Comment:
Is there a tracking ticket to make this change to were you planning to do that in this PR?

Reply:
Good point, created https://jira.mongodb.org/browse/CDRIVER-3682 to track this and updated the comment.